pkg/reconciler/events/auditlogs/auditlogs.go

```diff
@@ -0,0 +1,337 @@
+/*
+Copyright 2019 Google LLC.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auditlogs
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/google/knative-gcp/pkg/apis/events/v1alpha1"
+	listers "github.com/google/knative-gcp/pkg/client/listers/events/v1alpha1"
+	pubsubreconciler "github.com/google/knative-gcp/pkg/reconciler/pubsub"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+
+	"cloud.google.com/go/logging/logadmin"
+	glogadmin "github.com/google/knative-gcp/pkg/gclient/logging/logadmin"
+	gpubsub "github.com/google/knative-gcp/pkg/gclient/pubsub"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/cache"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/logging"
+)
+
+const (
+	finalizerName = controllerAgentName
+
+	resourceGroup = "stackdrivers.events.cloud.google.com"
+)
+
+type Reconciler struct {
+	*pubsubreconciler.PubSubBase
+
+	auditLogsSourceLister  listers.AuditLogsSourceLister
+	logadminClientProvider glogadmin.CreateFn
+	pubsubClientProvider   gpubsub.CreateFn
+}
+
+// Check that we implement the controller.Reconciler interface.
+var _ controller.Reconciler = (*Reconciler)(nil)
+
+// Reconcile compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the AuditLogsSource resource
+// with the current status of the resource.
+func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
```

Comment:
I remember when I tested the CAL source, the finalizer doesn't work. After I deleted the CAL source, the topic, the subscription and the logsink still exist in Google Cloud Platform, have you fixed this problem?

Reply:
The finalizer was implemented in [847257a739e6fce2d2880fc036220ac2af78d1af](https://github.com/google/knative-gcp/pull/462/commits/847257a739e6fce2d2880fc036220ac2af78d1af).